Update Python package build and upload example

jamescurtin opened the pull request:
### Why:

Closes https://github.com/github/docs/issues/6089

### What's being changed:

* Adds link to PyPA article on using Github Actions to publish packages to PyPI
* Adds link to instructions on how to create PyPI token (more secure than current example which uses username/password)
* Updates the example workflow to match https://github.com/actions/starter-workflows/pull/900 (same change being made in the `actions/starter-workflows` repo.)

### Check off the following:
- [x] I have reviewed my changes in staging. (look for the **deploy-to-heroku** link in your pull request, then click **View deployment**)
- [x] For content changes, I have reviewed the [localization checklist](https://github.com/github/docs/blob/main/contributing/localization-checklist.md)
- [x] For content changes, I have reviewed the [Content style guide for GitHub Docs](https://github.com/github/docs/blob/main/contributing/content-style-guide.md).

### Writer impact (This section is for GitHub staff members only):

- [ ] This pull request impacts the contribution experience
  - [ ] I have added the 'writer impact' label
  - [ ] I have added a description and/or a video demo of the changes below (eg. a "before and after video")

<!-- Description of the writer impact here -->

Commit:
Update Python package build and upload example

jamescurtin committed:
Merge branch 'main' into update-python-build-docs

ErdaKamarudin reviewed (changes requested):
Request change

janiceilene commented:
@jamescurtin Thanks so much for opening a PR! I'll get this triaged for review ⚡

Commit:
Merge branch 'main' into update-python-build-docs

sarahs commented:
👋 We did some content restructuring that resulted in a conflict with your PR. It looks like we may not have access to check out your fork. Let us know if you need a hand resolving the conflict!

Commit:
Merge remote-tracking branch 'upstream/main' into update-python-build-docs

jamescurtin commented:
> 👋 We did some content restructuring that resulted in a conflict with your PR. It looks like we may not have access to check out your fork. Let us know if you need a hand resolving the conflict!

Thanks for the ping! The conflict has been resolved.

Benzfccr7 reviewed (approved):
เพื่อนักลงทุน

Emmanuelchekwube commented:
I'm sure you will enjoy it

Commit:
Merge branch 'main' into update-python-build-docs

martin389 commented:
### Changes in this PR

| Article in staging  | Changes |
| --- | --- | 
“[Building and testing Python](https://docs-6090--update-python-build.herokuapp.com/en/actions/guides/building-and-testing-python)” | Instead of using `twine` to publish a package, this change uses an action called: `pypa/gh-action-pypi-publish`, and adds a recommendation for using using PyPI API tokens.

brcrista commented on content/actions/guides/building-and-testing-python.md:
```diff
@@ -421,14 +427,23 @@ jobs:
         run: |
-          python setup.py sdist bdist_wheel
-          twine upload dist/*
+          python -m build
```
For a one-line script, you don't need the `|`:

```
run: python -m build
```

martin389 replied:
Thanks, I've updated this now 👍

brcrista commented on content/actions/guides/building-and-testing-python.md:
```diff
@@ -421,14 +427,23 @@ jobs:
+        with:
+          user: __token__
+          password: ${{ secrets.TEST_PYPI_API_TOKEN }}
+          repository_url: https://test.pypi.org/legacy/
```
I'm not sure this is the right flow.  Pre-releases (e.g. v1.0.0-alpha1) should still go to production PyPI.  You would want to publish to test PyPI to smoke out issues with your release flow -- that's something you can do on every commit.

brcrista replied:
So I would just skip this step and run the next step without the `if`.

Benzfccr7 replied:
ผมทำไม่ค่อยเป็น

martin389 replied:
Thanks @brcrista - I've changed this now under this commit: https://github.com/github/docs/pull/6090/commits/fdb24c003a6fde54049f326d12d943b253e9a12b

brcrista commented on content/actions/guides/building-and-testing-python.md:
```diff
@@ -399,15 +399,21 @@ jobs:
 on:
   release:
-    types: [created]
+    types: [released, prereleased]
```
```suggestion
    types: [published]
```

If you save a *prerelease* as a draft and then publish it, you won't get the `prereleased` event.  This doesn't happen for non-pre releases; you just get `released` as usual.  If you just do `published`, you'll get everything.

https://docs.github.com/en/actions/reference/events-that-trigger-workflows#release

brcrista reviewed (commented):
Good improvement, thanks!  Left a few small comments

Commit:
Apply suggestions from code review

Co-authored-by: Brian Cristante <33549821+brcrista@users.noreply.github.com>

Commit:
Update building-and-testing-python.md

Commit:
Update building-and-testing-python.md

Commit:
Merge branch 'main' into update-python-build-docs

martin389 commented:
Thanks so much for the review @brcrista. I've added in the requested changes, if you're able to re-review 👍

Commit:
Update building-and-testing-python.md

jamescurtin commented:
Thanks for the updates, @martin389. FYI, I've also updated my PR for the corresponding action (https://github.com/actions/starter-workflows/pull/900) with these same requested changes.

Commit:
Update building-and-testing-python.md

Commit:
Update building-and-testing-python.md

brcrista reviewed (dismissed):
🚢

Commit:
Merge branch 'main' into update-python-build-docs

Commit:
Added small edits

Commit:
Update building-and-testing-python.md

Commit:
Update building-and-testing-python.md

Commit:
Merge branch 'main' into update-python-build-docs

martin389 commented:
Doing some final checks 👍

Commit:
Merge branch 'main' into update-python-build-docs

martin389 commented on content/actions/guides/building-and-testing-python.md:
```diff
@@ -397,17 +397,24 @@ jobs:
-You can store any access tokens or credentials needed to publish your package using secrets. The following example creates and publishes a package to PyPI using `twine` and `dist`. For more information, see "[Creating and using encrypted secrets](/github/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)."
+For this example, you will need to create two [PyPI API tokens](https://pypi.org/help/#apitoken). You can use secrets to store the access tokens or credentials needed to publish your package. For more information, see "[Creating and using encrypted secrets](/github/automating-your-workflow-with-github-actions/creating-and-using-encrypted-secrets)."
+
+The Python Packaging Authority (PyPA) also has its own [guide to using {% data variables.product.prodname_actions %}](https://packaging.python.org/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows).
```
I'm going to remove this link so that we can present consistent GH actions versions to the user, such as `checkout@v2`, `setup-python@v2`. But let me know if you think this is essential 👍

Commit:
Update building-and-testing-python.md

Removed link so that we can present consistent GH actions versions to the user, such as checkout@v2, setup-python@v2. But let me know if you think this is essential 👍

martin389 reviewed (approved):
Thanks for the awesome PR! 🎊 The dev team has approved this approach.

martin389 commented:
Changes are now live 🎊 
https://docs.github.com/en/actions/guides/building-and-testing-python#publishing-to-package-registries
